src/VisualQueryBuilder/components/OperationEditorBody.tsx

```diff
@@ -0,0 +1,288 @@
+import React, { useEffect, useRef, useState } from 'react';
+import { DraggableProvided } from 'react-beautiful-dnd';
+import { Button, Icon, Tooltip, useTheme2 } from '@grafana/ui';
+import { css, cx } from '@emotion/css';
+import { DataSourceApi, GrafanaTheme2, TimeRange } from '@grafana/data';
+import { OperationHeader } from './OperationHeader';
+import { QueryBuilderOperation, QueryBuilderOperationDefinition, QueryBuilderOperationParamDef, QueryBuilderOperationParamValue, VisualQuery, VisualQueryModeller } from '../types';
+import { getOperationParamEditor, getOperationParamId } from './OperationParamEditor';
+import { Stack } from '../../QueryEditor/Stack';
+import { v4 } from 'uuid';
+
+type Props = {
+    provided: DraggableProvided;
+    isConflicting: boolean,
+    index: number,
+    operation: QueryBuilderOperation,
+    definition: QueryBuilderOperationDefinition,
+    queryModeller: VisualQueryModeller;
+    query: VisualQuery;
+    onChange: (index: number, update: QueryBuilderOperation) => void;
+    onRemove: (index: number) => void;
+    onRunQuery: () => void;
+    datasource: DataSourceApi;
+    flash?: boolean,
+    highlight?: boolean,
+    timeRange?: TimeRange;
+  }
+  
+export function OperationEditorBody({ provided, flash, isConflicting, highlight, index, queryModeller, onChange, onRemove, operation, definition, query, timeRange, onRunQuery, datasource }: Props) {
+  const theme = useTheme2();
+  const styles = getStyles(theme, isConflicting);
+  const shouldFlash = useFlash(flash);
+  const { current: id } = useRef(v4());
+
+  const onParamValueChanged = (paramIdx: number, value: QueryBuilderOperationParamValue) => {
+    const update: QueryBuilderOperation = { ...operation, params: [...operation.params] };
+    update.params[paramIdx] = value;
+    callParamChangedThenOnChange(definition, update, index, paramIdx, onChange);
+  };
+
+  const onAddRestParam = () => {
+    const update: QueryBuilderOperation = { ...operation, params: [...operation.params, ''] };
+    callParamChangedThenOnChange(definition, update, index, operation.params.length, onChange);
+  };
+
+  const onRemoveRestParam = (paramIdx: number) => {
+    const update: QueryBuilderOperation = {
+      ...operation,
+      params: [...operation.params.slice(0, paramIdx), ...operation.params.slice(paramIdx + 1)],
+    };
+    callParamChangedThenOnChange(definition, update, index, paramIdx, onChange);
+  };
```

Comment:
I would conside `useCallback` these, now that we're at it, as it doesn't need to change the children on every render.